### PR TITLE
Allow login for old Twitter users with same email

### PR DESCRIPTION
**Pull request by josepegea**
One possible solution for #1063

As described in the issue mentioned above, with Twitter shutting down our authentication, users that only had that login configured are left out and cannot login. They can try to use different methods, but that means creating a new user.

One suggestion has been to add a new OAuth method based on OTP by email, which seems quite useful, but, still, wouldn't allow these users to connect to their old account.

As a special treatment for these special situation, this PR tries to allow these users to connect to their existing accounts if they log in using a different OAuth method and emails match.

The exact behaviour is:

1. If a new session successfully authenticates using a provider that includes email as part of their info.
2. If a user (and only one user) already exists with that same email
3. If that user was using Twitter (and only Twitter) as OAuth provider

If all of these conditions are met, then the new provider is added to the existing user and the login proceeds with that.

If the existing user had other providers already registered, then this logic doesn't kick in, as they can still use one of the others to log in.

There's still the problem with duplicate nicknames. They are taken from the "username" in the OAuth method, and if they are the same for different methods (for instance, the Twitter handle and the GitHub username are the same), even creating a new account will fail, because of the unique nicknames constrain. This could be a source of confusion and I think we should at least try to use a derived nickname (like "nickname_#{provider}") to prevent the duplication. But I think this should be solved in another PR.

Looking for comments and suggestions!

### CODE REVIEW

**Comment by phoet** [comment]
I'm fine with this. We should probably put an end of life to the implementation as it poses the theoretical the risk of account take-over.

**Comment by josepegea** [comment]
> I'm fine with this. We should probably put an end of life to the implementation as it poses the theoretical the risk of account take-over.

Good point! I added a deadline to the whole behaviour, defaulting to the end of this year, but configurable by an ENV var, if needed

File: app/models/authorization.rb [hunk 1]
```diff
@@ -14,9 +14,35 @@ def self.handle_authorization(existing_user, auth)
...
+
+  DEFAULT_TWITTER_USER_FALLBACK_DEADLINE = Date.new(2025, 1, 1)
+
+  def self.user_fallback_in_place?
+    deadline = begin
+      Date.parse(ENV['TWITTER_USER_FALLBACK_DEADLINE'])
+    rescue Date::Error, TypeError
+      DEFAULT_TWITTER_USER_FALLBACK_DEADLINE
+    end
+    Time.zone.today <= deadline
```
Comment (salzig) [inline app/models/authorization.rb:46]: Nitpick: 
```suggestion
    deadline.future?
```

**Review by salzig — APPROVED** [review]

**Review by JoschkaSchulz — APPROVED** [review]

**Comment by salzig** [comment]
FYI: around 25% of Twitter Accounts have a email address assigned. (in our database)

**Comment by josepegea** [comment]
True. The same happens to accounts in Madrid.rb.

I was thinking about how to solve that and the only idea that came to mind would be to:

- Let these users know about the situation
- Allow them to send a PM to the corresponding RUG Twitter account, providing their email address
- Have the RUG admin manually assign the email to the user, after validating that the Twitter handles match and that that account didn't have an email address before
-  Inform the users that they can log in using another provider that uses the same email (one option would be, once #1068 it's finished, email itself)

I know it's manual and not beautiful, but I can't think of anything else.

**Pull request merged**